### PR TITLE
Add 'exclusions' support to Azure FrontDoor's WebApplicationFirewall

### PR DESCRIPTION
**Pull request by wurp**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [x] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [x] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### CODE REVIEW

File: src/front-door/azext_front_door/_params.py [hunk 1]
```diff
@@ -181,6 +181,30 @@ def load_arguments(self, _):
         c.argument('policy_name', waf_policy_name_type, id_part=None)
         c.argument('rule_set_type', options_list=['--type'], help='Name of the ruleset with the overrides to list.')
 
+    with self.argument_context('network front-door waf-policy managed-rules exclusion add') as c:
+        c.argument('policy_name', waf_policy_name_type)
+        c.argument('rule_set_type', options_list=['--type'], help='Name of the ruleset to exclusion.')
```
[the remark this answers is not in the document]
Comment (wurp) [reply src/front-door/azext_front_door/_params.py]: The rule set is referred to by "type", as in "type of rules we're applying". This is used in the "az network front-door override" commands, which are already part of front-door CLI.

[the remark this answers is not in the document]
Comment (wurp) [reply src/front-door/azext_front_door/_params.py]: I will update it to say "id of".

File: src/front-door/azext_front_door/_params.py [hunk 2]
```diff
@@ -181,6 +181,30 @@ def load_arguments(self, _):
...
+        c.argument('rule_group_id', help='Name of the rule group containing the rule to exclusion.')
+        c.argument('rule_id', help='Name of the rule to exclusion.')
```
[the remark this answers is not in the document]
Comment (wurp) [reply src/front-door/azext_front_door/_params.py]: I will update it to say "id of"

File: src/front-door/azext_front_door/_params.py [hunk 3]
```diff
@@ -181,6 +181,30 @@ def load_arguments(self, _):
...
+        c.argument('match_variable', help='Which kind of variable\'s content will be ignored.')
+        c.argument('operator', help='Operator used to compare the variable name to the value.')
```
[the remark this answers is not in the document]
Comment (wurp) [reply src/front-door/azext_front_door/_params.py]: I will list example values here.

File: src/front-door/azext_front_door/_params.py [hunk 4]
```diff
@@ -181,6 +181,30 @@ def load_arguments(self, _):
         c.argument('policy_name', waf_policy_name_type, id_part=None)
         c.argument('rule_set_type', options_list=['--type'], help='Name of the ruleset with the overrides to list.')
 
+    with self.argument_context('network front-door waf-policy managed-rules exclusion add') as c:
+        c.argument('policy_name', waf_policy_name_type)
+        c.argument('rule_set_type', options_list=['--type'], help='Name of the ruleset to exclusion.')
+        c.argument('rule_group_id', help='Name of the rule group containing the rule to exclusion.')
```
[the remark this answers is not in the document]
Comment (wurp) [reply src/front-door/azext_front_door/_params.py]: this is the same as the existing help text for rule_group_id parameter for override commands.

File: src/front-door/azext_front_door/_params.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (MyronFanQiu) [reply src/front-door/azext_front_door/_params.py]: @wurp I see. But it's still confusing for me. Could we refine the help message for az network front-door override too?

[the remark this answers is not in the document]
Comment (MyronFanQiu) [reply src/front-door/azext_front_door/_params.py]: @wurp Any thoughts?

**Comment by wurp** [comment]
I agree; will do.

File: src/front-door/azext_front_door/_params.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (wurp) [reply src/front-door/azext_front_door/_params.py]: I agree; will fix them both.

**Comment by wurp** [comment]
What do we need next before we can "azdev extension publish front-door"? Just for this to be merged and released?

**Comment by MyronFanQiu** [comment]
@wurp Hello. You need to release this package by using `azdev extension publish`

**Comment by wurp** [comment]
@MyronFanQiu is there an ongoing incident with azure cli? When I try "az login --use-device-code", I can login through the web page as normal, but the command keeps hanging, eventually giving this response:
Max retries exceeded with url: /tenants?api-version=2016-06-01 (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x05416AB0>: Failed to establish a new connection: [WinError 10060] A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond',))

When I open portal.azure.com and try to open a terminal, it hangs and eventually errors out with "There was an error fetching tenants. Please reload the portal to try again."

**Comment by wurp** [comment]
> @MyronFanQiu is there an ongoing incident with azure cli? When I try "az login --use-device-code", I can login through the web page as normal, but the command keeps hanging, eventually giving this response:
> Max retries exceeded with url: /tenants?api-version=2016-06-01 (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x05416AB0>: Failed to establish a new connection: [WinError 10060] A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond',))
> 
> When I open portal.azure.com and try to open a terminal, it hangs and eventually errors out with "There was an error fetching tenants. Please reload the portal to try again."

Accessing https://management.azure.com/ from my developer desktop consistently times out. From my phone it loads OK.

**Comment by MyronFanQiu** [comment]
> > @MyronFanQiu is there an ongoing incident with azure cli? When I try "az login --use-device-code", I can login through the web page as normal, but the command keeps hanging, eventually giving this response:
> > Max retries exceeded with url: /tenants?api-version=2016-06-01 (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x05416AB0>: Failed to establish a new connection: [WinError 10060] A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond',))
> > When I open portal.azure.com and try to open a terminal, it hangs and eventually errors out with "There was an error fetching tenants. Please reload the portal to try again."
> 
> Accessing https://management.azure.com/ from my developer desktop consistently times out. From my phone it loads OK.

@jiasli Hello. Could you take a look? Do you have such login incident?

**Comment by wurp** [comment]
> 
> 
> > > @MyronFanQiu is there an ongoing incident with azure cli? When I try "az login --use-device-code", I can login through the web page as normal, but the command keeps hanging, eventually giving this response:
> > > Max retries exceeded with url: /tenants?api-version=2016-06-01 (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x05416AB0>: Failed to establish a new connection: [WinError 10060] A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond',))
> > > When I open portal.azure.com and try to open a terminal, it hangs and eventually errors out with "There was an error fetching tenants. Please reload the portal to try again."
> > 
> > 
> > Accessing https://management.azure.com/ from my developer desktop consistently times out. From my phone it loads OK.
> 
> @jiasli Hello. Could you take a look? Do you have such login incident?

Apparently it's pretty specific, maybe only to my machine. Other developers on my team aren't having this issue.

**Comment by wurp** [comment]
> 
> 
> > > > @MyronFanQiu is there an ongoing incident with azure cli? When I try "az login --use-device-code", I can login through the web page as normal, but the command keeps hanging, eventually giving this response:
> > > > Max retries exceeded with url: /tenants?api-version=2016-06-01 (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x05416AB0>: Failed to establish a new connection: [WinError 10060] A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond',))
> > > > When I open portal.azure.com and try to open a terminal, it hangs and eventually errors out with "There was an error fetching tenants. Please reload the portal to try again."
> > > 
> > > 
> > > Accessing https://management.azure.com/ from my developer desktop consistently times out. From my phone it loads OK.
> > 
> > 
> > @jiasli Hello. Could you take a look? Do you have such login incident?
> 
> Apparently it's pretty specific, maybe only to my machine. Other developers on my team aren't having this issue.

I've resolved my issue; it was local to my desktop and not due to any issue on the back end.

**Comment by MyronFanQiu** [comment]
@wurp 😄 good to know.

**Pull request merged**